### PR TITLE
Fixed min and max behavior for NaN values

### LINKED ISSUE #1366: behavior of cupy's min/max for NaN is different from numpy's
Cupy's min/max ignore NaN although numpy's returns NaN.

```
def check(xp):
    a = xp.array([np.NaN, -1, 0, 1])
    print a.min(), a.max()
check(np)
check(cupy)
```

output:
```
nan nan
1.0 -1.0
```

### PR DESCRIPTION
**Pull request by unnonouno**
Numpy's `min` and `max` return NaN when an array contains NaN value.
Note that `min` and `max` functions in CUDA ignore NaN values.
fix #1366

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1331,66 +1331,138 @@ divmod = create_ufunc(
...
+    if (a.index == -1) return b;
+    if (b.index == -1) return a;
+    return min_max_st<T>(min(a.value, b.value));
+}
+template <typename T>
+__device__ min_max_st<T> my_min_float(
+        const min_max_st<T>& a, const min_max_st<T>& b) {
+    if (a.index == -1) return b;
+    if (b.index == -1) return a;
+    if (isnan(a.value)) return a;
```
Comment (gwtnb) [inline cupy/core/core.pyx]: It seems that `isnan` is introduced in Visual Studio 2015. Is it okay? We can use `_isnan` for VS.

Comment (unnonouno) [inline cupy/core/core.pyx]: Oh, really? What's a recommended method to check nan?

Comment (gwtnb) [inline cupy/core/core.pyx]: AFAIK, the most portable way is `(x == x) != 0`

Comment (gwtnb) [inline cupy/core/core.pyx]: or wrapping `_isnan` on Windows.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1331,66 +1331,140 @@ divmod = create_ufunc(
...
     int index;
     __device__ min_max_st() : index(-1) { }
-    __device__ min_max_st(type_in0_raw v) : value(v), index(0) { }
-    __device__ min_max_st(type_in0_raw v, int i) : value(v), index(i) { }
+    __device__ min_max_st(T v) : value(v), index(0) { }
+    __device__ min_max_st(T v, int i) : value(v), index(i) { }
 };
-__device__ min_max_st my_min(const min_max_st& a, const min_max_st& b) {
+
+#define ISNAN(x) (x != x)
```
Comment (gwtnb) [inline cupy/core/core.pyx]: How about replacing with an inline (template) function?

Comment (gwtnb) [inline cupy/core/core.pyx]: This macro works for `float16` via [`operator float`](https://github.com/pfnet/chainer/blob/v1.13.0/cupy/core/carray.cuh#L35). It should be avoided to depend on such unobvious behavior. This behavior is an issue of cupy's float16, so it should be fixed in another PR.

**Comment by gwtnb** [comment]
LGTM

**Comment by unnonouno** [comment]
Do we need to fix `operator ==` behavior?

**Comment by gwtnb** [comment]
Yes, but in another PR.

**Comment by unnonouno** [comment]
OK
